src/app/modules/submission/encrypt-submission/encrypt-submission.middleware.ts

```diff
@@ -188,7 +190,84 @@ export const checkNewBoundaryEnabled = async (
 }
 
 /**
- * !! Do NOT enable `encryption-boundary-shift-virus-scanner` feature flag until this has been completely implemented. !!
+ * Helper function to trigger virus scanning and download clean file.
+ * @param response quarantined attachment response from storage submissions v2.1+.
+ * @returns modified response with content replaced with clean file buffer and answer replaced with filename.
+ */
+const triggerVirusScanThenDownloadCleanFileChain = (
+  response: ParsedClearAttachmentResponse,
+): ResultAsync<
+  ParsedClearAttachmentResponse,
+  VirusScanFailedError | DownloadCleanFileFailedError
+> =>
+  // Step 3: Trigger lambda to scan attachments.
+  triggerVirusScanning(response.answer)
+    .map((lambdaOutput) => lambdaOutput.body)
+    // Step 4: Retrieve attachments from the clean bucket.
+    .andThen((cleanAttachment) =>
+      // Retrieve attachment from clean bucket.
+      downloadCleanFile(
+        cleanAttachment.cleanFileKey,
+        cleanAttachment.destinationVersionId,
+      ).map((attachmentBuffer) => ({
+        ...response,
+        // Replace content with attachmentBuffer and answer with filename.
+        content: attachmentBuffer,
+        answer: response.filename,
+      })),
+    )
+
+/**
+ * Asynchronous virus scanning for storage submissions v2.1+. This is used for non-dev environments.
+ * @param responses all responses in the storage submissions v2.1+ request.
+ * @returns all responses with clean attachments and their filename populated for any attachment fields.
+ */
+const asyncVirusScanning = (
+  responses: ParsedClearFormFieldResponse[],
+): ResultAsync<
+  ParsedClearFormFieldResponse,
+  VirusScanFailedError | DownloadCleanFileFailedError
+>[] => {
+  return responses.map((response) => {
+    if (isQuarantinedAttachmentResponse(response)) {
+      return triggerVirusScanThenDownloadCleanFileChain(response)
+    }
+
+    // If field is not an attachment, return original response.
+    return okAsync(response)
+  })
+}
+
+/**
+ * Synchronous virus scanning for storage submissions v2.1+. This is used for dev environment.
+ * @param responses all responses in the storage submissions v2.1+ request.
+ * @returns all responses with clean attachments and their filename populated for any attachment fields.
+ */
+const syncVirusScanning = async (
```

Comment:
/minor suggestion to make it clear it is only for dev mode:
```suggestion
const devModeSyncVirusScanning = async (
```

Reply:
Okays thanks, done in [1f1edf0](https://github.com/opengovsg/FormSG/pull/6767/commits/1f1edf05c11b6bf4c44a4bf344c611f67b7cbd7d)!